app/javascript/components/tooltips/Concept.tsx

```diff
@@ -0,0 +1,244 @@
+    state.requestHover || state.requestFocus || state.hover || state.focus
+
+  switch (source) {
+    case 'request-hover':
+      nextState.requestHover = parsedAction && !hasOpenTooltip(id)
+      nextState.show = nextShow(nextState)
+      break
+    case 'request-focus':
+      nextState.requestFocus = parsedAction
+      nextState.show = nextShow(nextState)
+      break
+    case 'hover':
+      nextState.hover = parsedAction && !hasOpenTooltip(id)
+      nextState.show = nextShow(nextState)
+      break
+    case 'focus':
+      nextState.focus = parsedAction
+      nextState.show = nextShow(nextState)
+      break
+    case 'force-close':
+      nextState.show = false
+      break
+    default:
+      throw new Error('unhandled tooltip show action')
+  }
+
+  if (nextState.show) {
+    addOpenTooltip(id, dispatch)
+  } else {
+    removeOpenTooltip(id)
+  }
+
+  // request-focus and focus actions take precedence over mouse events
+  // so it can close any other for accessibility purposes
+  if (
+    (source === 'request-focus' || source === 'focus') &&
+    nextState.show &&
+    hasOpenTooltip(id)
+  ) {
+    closeOtherOpenTooltips(id)
+  }
+
+  return nextState
+}
+
+function dispatch(
+  id = '',
+  dispatcher: React.Dispatch<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
+  source: ShowActionSource,
+  action?: ShowActionValue
+): void {
+  dispatcher([id, dispatcher, source, action])
+}
+
+export const Concept = ({
+  endpoint,
+  parent,
+  hoverRequestToShow,
+  focusRequestToShow,
+  parentSlug = '',
+}: {
+  endpoint: string
+  parent: HTMLElement | null
+  hoverRequestToShow: boolean
+  focusRequestToShow: boolean
+  parentSlug?: string
+}): JSX.Element => {
+  const tooltipId = `concept-tooltip${
+    parentSlug ? `-${parentSlug}` : parentSlug
+  }`
+  // Retrieve the HTML contents from the endpoint
+  const { isLoading, isError, data: html_content } = useQuery(tooltipId, () =>
+    fetch(endpoint).then((res) => res.text())
+  ) as { isLoading: boolean; isError: boolean; data: string }
+
+  const tooltipRef = useRef(null)
+  const [{ show }, showDispatcher] = useReducer(showReducer, {
+    requestHover: false,
+    requestFocus: false,
+    hover: false,
+    focus: false,
+    show: false,
+  })
+
+  const popper = usePopper(parent, tooltipRef.current, {
+    placement: 'bottom',
+    modifiers: [{ name: 'offset', options: { offset: [0, 8] } }],
+  })
+
+  // For the next two useEffects:
+  // If I ignore the exhaustive deps warning and only have requestToShow, it
+  // works.  If I have the exhaustive deps added (popper) it causes inf loop
+  // due to popper reference changing on every render.
+
+  useEffect(() => {
+    if (hoverRequestToShow && popper?.update) {
+      popper.update()
+      dispatch(tooltipId, showDispatcher, 'request-hover', 'show')
+      return
+    }
+
+    setTimeout(
+      () => dispatch(tooltipId, showDispatcher, 'request-hover', 'hide'),
+      150
```

Comment:
this is the delay once the mouse leaves the concept before tool-tip disappears (unless it has another reason to stay open). It facilitates the ability to move the mouse across the gap.  I just picked a duration that felt good. Up for change.

Reply:
This is nice. Can you add this as a comment to the code please. The first two sentences of your comment here would be perfect 🙂 

Always good to add a comment for any "magic numbers" that aren't worth turning into a variable name (and I agree that here, inlining is best).

Reply:
Aha, that must be what's causing this:

![tooltip](https://user-images.githubusercontent.com/135246/99370499-0bb15380-28be-11eb-9d0a-b434a2b90ca0.gif)

Reply:
Ya, should just tweak the timing there a bit

Reply:
Thanks for the gif, makes it easier to see what's happening!

Reply:
First time making a GIF!

![Zvxx53q](https://user-images.githubusercontent.com/135246/99405050-990b9c80-28ec-11eb-9668-00794fb7ae4d.gif)

Reply:
This should now be fixed in https://github.com/exercism/v3-website/pull/273/commits/d803d7780035a47d9f9e87f7cd019873e43b8fd6